src/pairing.ts

```diff
@@ -0,0 +1,305 @@
+  ) {
+    this.randomFixLenVal = randomBytes(32, rng);
+    this.myCommittedStaticKey = commitPublicKey(this.myStaticKey.publicKey, this.randomFixLenVal);
+
+    if (pairingParameters instanceof InitiatorParameters) {
+      this.initiator = true;
+      this.qr = QR.fromString(pairingParameters.qrCode);
+      this.qrMessageNameTag = pairingParameters.qrMessageNameTag;
+    } else {
+      this.initiator = false;
+      this.qrMessageNameTag = randomBytes(MessageNametagLength, rng);
+      this.qr = new QR(
+        pairingParameters.applicationName,
+        pairingParameters.applicationVersion,
+        pairingParameters.shardId,
+        this.myEphemeralKey.publicKey,
+        this.myCommittedStaticKey
+      );
+    }
+    // We set the contentTopic from the content topic parameters exchanged in the QR
+    this.contentTopic = WakuPairing.toContentTopic(this.qr);
+
+    // Pre-handshake message
+    // <- eB {H(sB||r), contentTopicParams, messageNametag}
+    const preMessagePKs = [NoisePublicKey.fromPublicKey(this.qr.ephemeralKey)];
+
+    this.handshake = new Handshake({
+      hsPattern: NoiseHandshakePatterns.WakuPairing,
+      ephemeralKey: myEphemeralKey,
+      staticKey: myStaticKey,
+      prologue: this.qr.toByteArray(),
+      preMessagePKs,
+      initiator: this.initiator,
+    });
+  }
+
+  public getPairingInfo(): InitiatorParameters {
+    return new InitiatorParameters(this.qr.toString(), this.qrMessageNameTag);
+  }
+
+  public async getAuthCode(): Promise<string> {
+    return new Promise((resolve) => {
+      if (this.authCode) {
+        resolve(this.authCode);
+      } else {
+        this.eventEmitter.on("authCodeGenerated", (authCode: string) => {
+          this.authCode = authCode;
+          resolve(authCode);
+        });
+      }
+    });
+  }
+
+  public validateAuthCode(confirmed: boolean): void {
+    this.eventEmitter.emit("confirmAuthCode", confirmed);
+  }
+
+  private async isAuthCodeConfirmed(): Promise<boolean | undefined> {
+    // wait for user to confirm or not, or for the whole pairing process to time out
+    const p1 = pEvent(this.eventEmitter, "confirmAuthCode");
+    const p2 = pEvent(this.eventEmitter, "pairingTimeout");
+    return Promise.race<boolean | undefined>([p1, p2]);
+  }
+
+  private async initiatorHandshake(): Promise<[NoiseSecureTransferEncoder, NoiseSecureTransferDecoder]> {
+    // The handshake initiator writes a Waku2 payload v2 containing the handshake message
+    // and the (encrypted) transport message
+    // The message is sent with a messageNametag equal to the one received through the QR code
+    let step = this.handshake.stepHandshake({
+      transportMessage: this.myCommittedStaticKey,
+      messageNametag: this.qrMessageNameTag,
+    });
+
+    // subscribe to new content topic where reply to first handshake message will arrive
+    const hsContentTopic = WakuPairing.toCustomContentTopic(this.qr, "TODO-HSContentTopic"); // TODO: how to calculate ct-id?
```

Comment:
@s1fr0 this is the content topic used to reply to the 1st handshake message. You suggested that this content topic should be deterministically derived from the content of the first handshake message. Should I use a hkdf of the transport message? Also, do you have an example of the format that should be used for this content topic?